plugins/ccgx/fu-ccgx-i2c.c

```diff
@@ -0,0 +1,774 @@
+			     error_local->message);
+
+		return FALSE;
+	}
+	return TRUE;
+
+}
+
+static gboolean
+wait_for_notification (FuDevice *device, CyI2CDeviceHandle* handle, guint16 *bytes_pending, guint32 io_timeout, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+	GUsbDevice *usb_device = NULL;
+	gsize actual_length = 0;
+	guint8 ep_num;
+	guint8 i2c_status[CY_I2C_EVENT_NOTIFICATION_LEN] = {0};
+
+	g_return_val_if_fail (handle != NULL, FALSE);
+	g_return_val_if_fail (FU_USB_DEVICE (device), FALSE);
+	usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	g_return_val_if_fail (usb_device != NULL, FALSE);
+
+	ep_num = handle->ep.intr_in;
+
+	if (g_usb_device_interrupt_transfer (usb_device,
+					     ep_num,
+					     (guint8*)i2c_status,
+					     CY_I2C_EVENT_NOTIFICATION_LEN,
+					     (gsize*)&actual_length,
+					     io_timeout,
+					     NULL,
+					     &error_local)) {
+
+		if (i2c_status[0] & CY_I2C_ERROR_BIT) {
+			if (i2c_status[0] & 0x80) { /* write */
+				if (i2c_reset (device,handle,CY_I2C_MODE_WRITE,NULL)  == FALSE) {
+					g_warning ("failed to reset i2c for write while getting i2c event");
+				}
+				memcpy(bytes_pending, &i2c_status[1], 2);
+
+			} else { /* read */
+				if (i2c_reset (device,handle,CY_I2C_MODE_READ,NULL) == FALSE) {
+					g_warning ("failed to reset i2c for read while getting i2c event");
+				}
+				memcpy(bytes_pending, &i2c_status[1], 2);
+			}
+			g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "i2c status error in i2c event : 0x%x", 
+			     (guint8) i2c_status[0]);
+
+			return FALSE;
+		}
+	} else {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "failed to get i2c event:%s", 
+			     error_local->message);
+		if (!g_usb_device_reset (usb_device, NULL)) {
+			g_warning ("failed to reset i2c while getting i2c event");
+		}
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_i2c_read:
+ * @device: #FuDevice
+ * @handle: I2C handle
+ * @data_cfg  I2C data configuration
+ * @data_buffer  I2C data buffer
+ * @error: a #GError or %NULL
+ *
+ * Read	data through I2C
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_i2c_read (FuDevice *device, CyI2CDeviceHandle *handle, CyI2CDataConfig *data_cfg, CyDataBuffer *data_buffer, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+	guint16 scb_index = 0;
+	guint32 io_timeout = FU_CCGX_I2C_WAIT_TIMEOUT;
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	guint16 w_value, w_index, w_length;
+	guint8 bm_request;
+	gsize actual_length = 0;
+	guint8 ep_num;
+	guint8 i2c_status[CY_I2C_EVENT_NOTIFICATION_LEN] = {0};
+	guint8 mode = CY_I2C_MODE_READ;
+	guint8 slave_address = 0;
+	guint16 byte_pending = 0;
+	guint64 elapsed_time = 0;
+	g_autoptr(GTimer) start_time = g_timer_new ();
+
+	if (!i2c_get_status (device, handle, mode, (guint8*)i2c_status, error)) {
+		g_prefix_error (error, "i2c read error: ");
+		return FALSE;
+	}
```

Comment:
i2c_status doesn't seem to be read -- is the value important?

Reply:
I need to check i2c status error here but error checking code is missed.
I will add i2c status error checking code here.